### PR TITLE
feat(review): classifyThreads decision table + planReviewBatch (E2)

### PR DESCRIPTION
**Pull request by camerontaylor**
## Goal
E2 — `classifyThreads` + `planReviewBatch`. Breakdown: ws-e scope items 2–3; plan §5 op table (review ops); UC §2 rows 34–35; invariant I6.

## Stack
parent: `merge-queue` (E1 merged + promoted; this branch is cut from main @ 5c07c66); children: `p2-ef/e3-review-verify` (next in stack)

## Claims (acceptance checks this PR satisfies)
- [x] ws-e: "classifyThreads is a pure decision table with full verdict coverage; injected `nowMs`" — 15-row decision table (`src/ops/review/classifyThreads.ts`, module doc), 34 tests with a covering test per row + precedence + null-timestamp/null-author fallbacks (`test/ops/review/classifyThreads.test.ts`); `nowMs` is a required parameter and the only clock (no `Date.now()` anywhere).
- [x] ws-e: "planReviewBatch … per-item isolation is the asserted default" (I6) — `planReviewBatch.ts` defaults to isolated one-item batches (asserted in test "I6 default: a single actionable item → exactly ONE isolated batch…"); shared-worktree fan-out is a supported option with path grouping and a hard `maxItemsPerSharedBatch` cap.
- [x] Verdict vocabulary frozen and exported for WS-J/WS-F: `ThreadVerdict` = `actionable | responded | resolved | blocked | skip`, re-exported from the family barrel `src/ops/review/index.ts`.

## Deferred to
- Verdict-set consumption by the review-classifier eval role → WS-J (phase-2 lane j, fixtures repo)
- Batch dispatch to actual fixer workers → phase-3 lane E (E4 fixReviewItem)
- `review-loop` plan wiring (ws-e item 8) → phase-3 lane E, goal E4 (same disposition as PR 63, audit-confirmed)

## Invariants exercised
I6: per-item isolation is the asserted default (`planReviewBatch` tests, named for I6); shared mode caps batch size and is opt-in only.
I2 (vocabulary half): bot skip/failure notices are skip-patterns in `classify.config.ts` (not reviews); no reviewer privileged — responder = PR author per config data.
I11: untouched by this PR (no new GH facts); I1/I3 not touched.

## Review rounds
round 1: 7 findings (2 med, 5 low) → all 7 fixed in c8d68a6
round 2: 6 findings (2 med, 4 low) → all 6 fixed in eaa18b1
round 3: 5 findings (2 med, 3 low) → all 5 fixed post-cap in 8b31c29; **cap reached (3 rounds), no fourth round — best-effort residue: post-cap fixes gate-verified but not re-reviewed**
Accepted at merge (best effort): the round-3 + CodeRabbit fixes (98c50ae) are gate-verified but not re-reviewed by a fresh reviewer.

## Review threads (bots)
2 coderabbitai threads: 2 fixed (commits c8d68a6 — pattern anchoring; 98c50ae — stateless lastIndex evaluation; replies name the commits) / 0 dismissed / 0 deferred; auto-resolved by CodeRabbit with changes verified present, replies posted per reply-before-resolve discipline; 0 unresolved at the pre-merge query.
**POST-MERGE RACE (recorded deviation):** a third thread (chatgpt-codex-connector, `classify.config.ts` — the configuration-error skip pattern lacks a bot-identity anchor) arrived between the final pre-merge query and the merge; the merge executed with it unresolved, violating the zero-unresolved precondition. Remediation: the fix lands as the first commit of the immediately-following lane PR (p2-ef/e3-review-verify), after which the thread gets its reply (naming that commit) and is resolved. Reported to the conductor as a process deviation.

## Dismissed findings
none yet

## Deviation
- (traceability, recorded for the auditor) The brief's "note for lane j" says the root `src/index.ts` barrel "already re-exports" the review family — it does not yet (lane i owns the barrel; its PR is still open in this phase). The vocabulary is exported from the family barrel `src/ops/review/index.ts`; root-barrel re-export lands via lane i. Recorded here and in the lane STATUS rather than silently absorbed (README §7).

## Verification
At `8b31c29` in the lane worktree: `npm run typecheck` OK · `npm run lint` OK · `npm test` 29 files, 651 passed / 7 pre-existing skips · `npm run denylist` clean. Purity: no I/O, no wall clock — `nowMs` the only clock (injected).

### CODE REVIEW

**Comment by camerontaylor** [comment]
## Fresh code review — round 1 (FINDINGS: 7)

Fresh reviewer agent (claude-zai/glm-5.3, round 1), posted by the lane leader (no-credential deviation, §3.8). Reviewer verdict: purity holds (nowMs proven to drive verdicts), row precedence pinned, I6 isolated default asserted by name, I2 holds, no injection/ReDoS surface; findings concentrate in the review table and one dropped upstream contract.

1. **[medium]** No responder-authored row in the review table (`classifyThreads.ts:253`) — the PR author's own comment-reviews classify `actionable` → fixer batch for the responder's own words, inconsistent with rows 2/10. → **fixed** (new row, gated on `skipResponderAuthoredThreads`, + test)
2. **[medium]** `FetchedReviewState.truncated`/`truncatedBecause` silently dropped (`classifyThreads.ts:309`) — fetchReviewState's doc says every consumer MUST consult the flag; a `reviewThreads.lag` fetch means a fresh thread is missing and never gets a verdict. → **fixed** (classifyThreads now returns `Classification = { items, truncated, truncatedBecause }`; planReviewBatch consumes `items`; module doc states the contract)
3. **[low]** `review.state` never read — DISMISSED reviews classify `actionable`. → **fixed** (skip row + `skipDismissedReviews` config data, default true, + test)
4. **[low]** `\breview\b.*\bfailed\b` matches human phrasing → genuine feedback silently skipped. → **fixed** (pattern tightened to bot-anchored phrasing as R3 data)
5. **[low]** Under `treatNullCreatedAtAs: 'nowMs'` a null-timestamp responder reply auto-answers every past review (fail-open), contradicting the config doc. → **fixed** (row 8 requires a real parsed reply timestamp; null/unparseable reply timestamps never count as answering; config doc now matches behavior; test flipped)
6. **[low]** Row 8 scans only `restReviewComments` — top-level issue comments ("done in <sha>") never mark a review answered. → **fixed** (row 8 scans both `restReviewComments` and `restIssueComments`; documented)
7. **[low]** The R3-tunable config flips (`blockOnOutdatedThreads`, `skipResponderAuthoredThreads`) have zero test coverage. → **fixed** (`test.each` both-values pairs mirroring the timestamp pairs)

Disposition: 2 medium + 5 low, all fixed in the round-1 fix commit; nothing dismissed, nothing recorded-unfixed.

**Comment by camerontaylor** [comment]
Round-1 findings addressed in commit c8d68a6: F1 responder-authored review row (responder_authored, skipResponderAuthoredThreads-gated, precedence pinned), F2 truncation propagated (classifyThreads returns Classification {items, truncated, truncatedBecause}; consumers MUST consult before dispatch — module doc + tests), F3 DISMISSED-review skip row + skipDismissedReviews config data (default true, both values tested), F4 bot-anchored failure patterns (human "review failed to consider" stays actionable — tested), F5 row-10 requires a real parsed reply timestamp (null/unparseable never answers; config doc corrected; old test flipped), F6 answer scan covers restReviewComments AND restIssueComments (tested), F7 both-values coverage for the R3-tunable flags. Table now 14 rows, documented with ordering rationale. Gates at c8d68a6: typecheck/lint/denylist clean, 624 passed / 7 pre-existing skips.

File: src/ops/review/classify.config.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+ * literal mirrors them so the shipped values are greppable in one place).
+ */
+export const defaultClassifyConfig: ClassifyConfig = {
+  // Bot skip/failure notices are not reviews (I2): each pattern matches a
+  // distinct real-world phrasing class. R3 refines AS DATA.
+  skipPatterns: [
+    // "CodeRabbit ... skipped ..." — the bot punted on this PR.
+    /\bCodeRabbit\b.*\bskipped\b/i,
+    // "review ... failed" — a review tool errored out.
+    /\breview\b.*\bfailed\b/i,
```
[the remark this answers is not in the document]
Comment (camerontaylor) [reply src/ops/review/classify.config.ts]: Fixed (your proposed direction, stronger form) in commit c8d68a6: the default failure pattern is now bot-anchored — /(?:CodeRabbit|Codex|coderabbitai|chatgpt-codex-connector)…(?:failed|error)/ — so human phrasing like "The review endpoint failed for null input" classifies actionable, not skip. Pattern stays config DATA per the module contract. Test pins the human-sentence case.

File: src/ops/review/classifyThreads.ts [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+  return config.treatNullCreatedAtAs === 'nowMs' ? nowMs : 0;
+};
+
+/** True when the author is the responder. Null authors are NEVER the responder. */
+const isResponder = (authorLogin: string | null, responder: string | null): boolean =>
+  authorLogin !== null && authorLogin === responder;
+
+/** True when the body matches any configured bot skip/failure pattern (I2). */
+const matchesSkipPattern = (body: string, config: ClassifyConfig): boolean =>
+  config.skipPatterns.some((pattern) => pattern.test(body));
```
[the remark this answers is not in the document]
Comment (camerontaylor) [reply src/ops/review/classifyThreads.ts]: Fixed in commit 98c50ae: matchesSkipPattern now evaluates each config pattern via a fresh expression with g/y flags stripped, so lastIndex state can never leak between calls; regression tests pin a shared /g pattern and a /y pattern across three consecutive classifications.

**Comment by camerontaylor** [comment]
## Fresh code review — round 2 (FINDINGS: 6)

Fresh reviewer agent (claude-zai/glm-5.3, round 2 — new process), posted by the lane leader (no-credential deviation, §3.8). Reviewer verified round-1/CodeRabbit fixes genuinely present; I2/I6 hold; purity verified.

1. **[medium]** No row for APPROVED / empty-body / null-state reviews (`classifyThreads.ts:339`) — an external "Ship it" approval falls through to `actionable` and plans a fixer batch for non-feedback, contradicting the vocabulary's own definitions. → **fixed** (new row: APPROVED with empty/whitespace body → skip `approval_no_body`; null-state with empty body → skip; bodies with text stay actionable — fails toward action; config field `skipApprovalReviews` default true; + tests)
2. **[medium]** `\bskipping review\b` / `\bnot reviewing\b` unanchored, violating the file's own MUST-anchor rule (`classify.config.ts:100-102`) — a human's "I'm not reviewing the migrations this pass, but …" would skip the whole body. → **fixed** (both patterns anchored on bot/tool identity per the MUST; + human-phrase tests)
3. **[low]** `ClassifyConfig.responderIs` inert (never read; doc claims R3 can widen AS DATA). → **fixed** (RowContext switches on `config.responderIs`; 'pr-author' → state.authorLogin; widening later is now genuinely data-only)
4. **[low]** `truncatedBecause` aliases the input state's array (`classifyThreads.ts:425`). → **fixed** (copied)
5. **[low]** A null-timestamp responder reply wins the last-word check under the nowMs fallback → `responded` (fail-open; asymmetry with row 10's parseRealMs rule). → **fixed** (the responder-last-word row now requires a real parsed timestamp on the last reply; null-timestamp replies fall through to actionable; TN1 test flipped + doc updated)
6. **[low]** The MUST-consult-truncated contract is prose-only (`planReviewBatch.ts:110`) — nothing structural forces dispatch to check. → **fixed** (planReviewBatch now takes the `Classification` and fails closed — throws naming `truncatedBecause` — when `truncated === true`; tests updated)

Also corrected: PR body said "12-row decision table"; the table is 14 rows (body updated). Disposition: 2 medium + 4 low, all fixed; nothing dismissed, nothing recorded-unfixed.

**Comment by camerontaylor** [comment]
Round-2 findings addressed in commit eaa18b1: R2-1 approval/empty-summary row 11 (APPROVED+empty → approval_no_body; null-state+empty → empty_summary_no_state; skipApprovalReviews config, non-empty bodies stay actionable), R2-2 self-skip patterns anchored on bot identity per the MUST (human "not reviewing … but" stays actionable — tested), R2-3 responderIs now read via exhaustive responderOf switch, R2-4 truncatedBecause defensively copied, R2-5 responder-last-word requires a real parsed timestamp (null falls to actionable; TN1 flipped), R2-6 planReviewBatch now takes the full Classification and throws on truncated=true (structural enforcement). Table is now 15 rows; module doc, config docs, and tests agree. Gates at eaa18b1: typecheck/lint/denylist clean, 642 passed / 7 pre-existing skips.

**Comment by camerontaylor** [comment]
## Fresh code review — round 3 (FINDINGS: 5)

Fresh reviewer agent (claude-zai/glm-5.3, round 3 — new process), posted by the lane leader (no-credential deviation, §3.8). Reviewer verified purity, truncation refusal, I6 default, stateless pattern evaluation, null-author externality, and E1 type/field agreement; three suspected defects confirmed by executing the module against minimal fixtures. This is the third and final review round (protocol cap); dispositions below are post-cap fixes.

1. **[medium]** Empty-body `COMMENTED` reviews (GitHub's standard inline-only review wrapper) fall through to `actionable` → one phantom fixer batch per inline review under the I6 default (`classifyThreads.ts:380`). → **fixed post-round** (row 11 gains the COMMENTED arm, reason `commented_no_body`; empty CHANGES_REQUESTED stays actionable — the verdict itself is signal, documented; + tests)
2. **[medium]** Row 10's answer scan unscoped: any responder comment postdating a review marks it answered, so one late thread-reply voids every older unaddressed summary (fail-open). → **fixed post-round** (summary-answer evidence restricted to top-level issue comments; thread replies answer threads via row 5 only; + multi-round test where review A stays actionable)
3. **[low]** Pattern 2 lacks the trailing `\b` after the identity alternation — "Codex-style tooling failed us" prefix-matched → skip. → **fixed post-round** (boundary added; + test)
4. **[low]** Thread-table cross-verdict precedence pairs 3>4 and 4>5 unpinned. → **fixed post-round** (two precedence tests)
5. **[low]** Test-file banner says "rows 1–14" (actual 15); PR body repeated the stale count. → **fixed post-round** (banner; PR body corrected to 15-row by the leader)

**Best-effort residue (recorded):** the reviewer loop ended at the 3-round cap with a non-zero round. The 5 findings are fixed in the post-cap commit and re-gated, but NOT re-reviewed by a fresh reviewer — the spec auditor verifies conformance.

**Comment by camerontaylor** [comment]
Round-3 findings fixed post-cap in commit 8b31c29 (no fourth review round — cap honored): R3-1 row 11 COMMENTED arm (commented_no_body; empty CHANGES_REQUESTED stays actionable — documented), R3-2 summary-answer evidence restricted to top-level issue comments only (thread replies answer threads via row 5; multi-round test pins older unaddressed review stays actionable), R3-3 pattern-2 trailing boundary (+ (?!-) guard — the literal suggested form still prefix-matched, verified empirically), R3-4 thread precedence pairs 3>4 and 4>5 pinned, R3-5 banner/count alignment (15 rows everywhere). Gates at 8b31c29: typecheck/lint/denylist clean, 651 passed / 7 pre-existing skips.

**Comment by camerontaylor** [comment]
## Spec-conformance audit — round 1

Fresh read-only auditor (claude-zai/glm-5.3, plan mode), posted by the lane leader. **VERDICT: ALL-PASS** — every row PASS (or correctly-untouched/DEFERRED with a confirmed owner). Highlights: the five-verdict vocabulary is a code-frozen TS union outside the config (R3 can tune values as data, never absorb vocabulary — README §7 rule verified structurally); I2's truncation-fails-closed half is implemented end-to-end (verbatim propagation + structural refusal in planReviewBatch); all three deferral rows have confirmed later-goal owners (WS-J lane-j:14-15, phase-3-lane-e.md E4 ×2); the root-barrel deviation is recorded in the PR body, not absorbed; review threads 0 unresolved with the two CodeRabbit fixes verified on-branch. Auditor notes (non-blocking): PR-body test count is stale (34 → actual 82 cases at HEAD — understates coverage); STATUS.md deviation entry lands post-merge per protocol.

**Comment by camerontaylor** [comment]
all-clear: review r1 7/r2 6/r3 5 findings (cap at 3 rounds; r3 + CodeRabbit fixes gate-verified post-cap, best-effort recorded), audit round 1 ALL-PASS, threads: 2 fixed / 0 dismissed / 0 deferred → 0 unresolved (verified pre-merge). CI green at 8b31c29 (8 checks). Merging.

File: src/ops/review/classify.config.ts [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    // A known bot/tool identity at LINE START followed within one line by
+    // "failed"/"error" — the tool's own failure notice. The trailing \b
+    // after the identity mirrors the `\bCodeRabbit\b` shape, and the
+    // `(?!-)` guard keeps a HYPHENATED tool-name MENTION from reading as
+    // the tool speaking ("Codex-style tooling failed us here — please fix
+    // the harness manually." is a human's sentence; \b alone is satisfied
+    // before the hyphen and would eat it).
+    /^\s*(?:(?:CodeRabbit|coderabbitai|Codex|chatgpt-codex-connector)\b)(?!-)[^\n]{0,80}\b(?:failed|error)\b/i,
+    // Tooling self-skip caused by a configuration/setup problem.
+    /\b(?:configuration|setup)\s+(?:error|problem)[^\n]{0,40}\bskipping\b/i,
```
[the remark this answers is not in the document]
Comment (camerontaylor) [reply src/ops/review/classify.config.ts]: Fixed in commit e610c60 on p2-ef/e3-review-verify (the immediate follow-up PR): the configuration-error skip pattern is rebuilt in the same line-start bot-identity-anchored form as its siblings — `^\s*(?:(?:CodeRabbit|coderabbitai|Codex|chatgpt-codex-connector)\b)(?!-)…(?:configuration|setup)\s+(?:error|problem)…\bskipping\b`. Your exact example sentence is pinned in the test suite as actionable (never `bot_skip_notice`), and a genuine bot configuration-error notice pins the skip direction. Process note: this thread arrived in the race window before PR 71 merged and could not be fixed inside that PR; the zero-unresolved precondition violation is recorded in the PR-71 body and reported to the conductor.
